Checkpoint 2

andela-joyebanji opened the pull request.

Commit:
Initial commit

Commit:
[Chore] Remove testing files

- Delete indexPDO.php
- Delete potatoTest.php
- Delete test.php

Commit:
[Chore] Remove unnecessary __callStatic method

Commit:
[Feature] Implement delete operation in model

Commit:
[Chore] Refactor Save Operation

- Change how querys are constructed in update and create method
- Fix little bugs in update and create method

Commit:
[Chore] Update comments

Update the comments in various methods in order to maintain consistency.
Add missing ; in update and create

Commit:
[Chore] Refactor codebase

Add config property in DatabaseConnection for storage of configuration values.
Add missing semi-colon in Model class.

[Chore] Refactor codebase

Commit:
[Chore] Refactor codebase

Add id property annotation
Remove unused variables

Commit:
Add travis integration

Commit:
[Chore] Add new tests

Add test cases in DatabaseConnectionStringFactoryTest
Fix namespacing issue in DatabaseConnectionStringFactory and DatabaseDriverNotSupportedException class

[Chore] Add new tests

Add DatabaseConnectionTest file to test the DatabaseConnection class
Add HelperTest file to test the Helpers class
Change config variables
Commit destroy test action in index.php

[Chore] Add new test cases

Add new test case for getInstance, setDefaultOptions, and getDefaultOptions

[Chore] Add new test cases

Add Reflection method to test class with private methods
Add test cases for TryAgainIfCausedByLostConnection method

Commit:
[Fix] Fix logical mistakes and add new tests

Fix logical errors in class Model
Add new file, ModelTest, for testing the Model class
Modified config file

Commit:
[Feature] Make DatabaseConnection implement a interface

In order to make the DatabaseConnection substituable with other class, DatabaseConnection class
now implement the DatabaseConnectionInterface
Modify config to in sqlite connection
Change method name in test

Commit:
[Chore] Add new tests

Add \ to the extended Exception class in other to make is reference the global Exception class
Add new methods and comments in Model class in other to make the class more testable and concise
Add new test case in ModelTest class
Fix typo in DatabaseConnectionTest class

[Chore] Add new tests

Add tests cases for save and Magic methods

[Chore] Add new test case

Add test case for the getProperties method

Commit:
[Chore] Remove unnecessary comments

Commit:
[Chore] Applied fixes from StyleCI

Commit:
[Chore] Applied fixes from StyleCI

Commit:
[Chore] Update README file

Commit:
[Chore] Update README file

Commit:
[Chore] Update README.md

Fix Markdown issues

Update README.md

Commit:
[Feature] Integrate Infector

Infector is a Doctrine package used to pluralize words. It's used to pluralize model names.
The Model class now use Infector.

Commit:
[Chore] Applied fixes from StyleCI

Commit:
[Feature] Integrate Dotenv

Include the Dotenv package in the composer.json file
Modify DatabaseConnection file to make use of env varibles loaded using Dotenv package

Commit:
[Fix] Change env variables lookup array

Commit:
[Chore] Applied fixes from StyleCI

Commit:
[Chore] Added new test cases

Write new test cases on DatabaseConnectionStringFactory and DatabaseConnection class

Commit:
[Chore] Applied fixes from StyleCI

Commit:
[Chore] Update README.md

Commit:
[Chore] Add new test case

Commit:
[Chore] Rename test case

Commit:
[Chore] Refactor Codebase

Remove unused variables in DatabaseConnectionTest file
Remove unnecessary comments in ModelTest file
Add new comments in ModelTest file
Update comment in Model class file

Commit:
Applied fixes from StyleCI

Commit:
Applied fixes from StyleCI

Commit:
Merge pull request #9 from andela-joyebanji/analysis-qorGYP

[Chore] Applied fixes from StyleCI

andela-anandaa commented on src/DatabaseConnection.php:
```diff
@@ -0,0 +1,175 @@
+      */
+     public function loadEnv()
+     {
+         if (!getenv('APP_ENV')) {
```
**Awesome:** Best practice, making use of `.env` files for loading configurations.

andela-anandaa commented on src/Model.php:
```diff
@@ -0,0 +1,241 @@
+     */
+    public function __set($key, $val)
+    {
+        $this->properties[$key] = $val;
```
**Nitpick:** What if the key is not available?

andela-anandaa replied:
@andela-joyebanji - What this fixed?

andela-joyebanji replied:
Yes. It was fixed.

andela-anandaa commented on src/Model.php:
```diff
@@ -0,0 +1,241 @@
+     */
+    public function getTableName()
+    {
+        $className = explode('\\', get_called_class());
```
**Nitpick:** `end` should be brought here for semantics: `end(explode('\\', get_called_class())`

andela-joyebanji replied:
I tried your suggested way but I got an error `Only variables should be passed by reference`. This is because the end function requires an argument that can be passed by reference. The means the return value of `explode('\\', get_called_class())` must be stored in a variable before it can then be passed to the `end` function.
Reference: http://php.net/end

andela-anandaa commented on README.md:
```diff
@@ -1,2 +1,128 @@
-# PotatoORM
-A simple agnostic ORM that can perform the basic crud database operations.
```
**Nitpick:** "CRUD" should be in capitals. Will also be good to give some context about this -- a _Checkpoint 2_ for Andela Fellowship.

andela-anandaa commented on README.md:
```diff
@@ -1,2 +1,128 @@
+    2. Postgres 
+    3. SQLite
+    
+You also need set your environment variables to define your database parameters or create a _.env_ file in your project.
```
Use back-tick instead when quoting a file, i.e. `.env` file instead of _.env_.

andela-anandaa commented on README.md:
```diff
@@ -1,2 +1,128 @@
+
+### Exceptions
+
+	DatabaseDriverNotSupportedException: This exception is thrown when database driver in not supported.
```
This can be bullets.

Commit:
[Fix] Fix issues raised during review

Update README
Table name can be specified in Models
Check if the requested value exist before returning a value in __get magic method
Remove redundant id value in test array in ModelTest

Commit:
Merge branch 'develop' of https://github.com/andela-joyebanji/PotatoORM into develop

Commit:
Applied fixes from StyleCI

Commit:
Applied fixes from StyleCI

Commit:
Merge pull request #10 from andela-joyebanji/analysis-864llN

[Chore] Applied fixes from StyleCI

Commit:
[Fix] Fix issue raised during review

Get now returns null when a Model is not found
Update test case for Get when a model is not found

Commit:
[Chore] Merge branch 'develop' of https://github.com/andela-joyebanji/PotatoORM into develop

Commit:
Applied fixes from StyleCI

Commit:
Merge pull request #11 from andela-joyebanji/analysis-q2Q19Z

[Chore] Applied fixes from StyleCI

andela-anandaa commented:
#### README
- Well written README, just fix the minor nitpicks noted above.
- Good: badges for _QuantifiedCode, Travis and Coveralls_.
- Add a link to the live application on the README.

#### Github Commits
- The commit messages are okay and consistent. However, next time, make sure all the commit messages have at least a body that explains the _what_ and _why_.

#### Comments
- The code is well commented.

#### Code Quality
- Good: good OOP design patterns.
- Read more [here](http://culttt.com/2014/04/09/use-exception/) about when you should use exceptions.

#### Functionality
- Had issues with object type being returned, _stdClass_ instead of the model class created by the end user -- but fixed on review.
- Good: Code works as expected.

#### Testing
- Well structured tests.
- Good: 91% code coverage.

Commit:
[Fix] Amend issues observed during code review

GetAll now return a Model class object
Add test case for __get of unset property

Commit:
[Chore] Merge branch 'develop' of https://github.com/andela-joyebanji/PotatoORM into develop

Commit:
Applied fixes from StyleCI

Commit:
Merge pull request #12 from andela-joyebanji/analysis-ze72E9

[Chore] Applied fixes from StyleCI

Commit:
Remove ModelNotFoundException file
